internal/controllers/cluster/cluster_controller_phases.go

```diff
@@ -143,132 +141,149 @@ func (r *Reconciler) reconcileExternal(ctx context.Context, cluster *clusterv1.C
 		)
 	}
 
-	return external.ReconcileOutput{Result: obj}, nil
+	return obj, nil
 }
 
 // reconcileInfrastructure reconciles the Spec.InfrastructureRef object on a Cluster.
-func (r *Reconciler) reconcileInfrastructure(ctx context.Context, cluster *clusterv1.Cluster) (ctrl.Result, error) {
+func (r *Reconciler) reconcileInfrastructure(ctx context.Context, s *scope) (ctrl.Result, error) {
 	log := ctrl.LoggerFrom(ctx)
-
-	if cluster.Spec.InfrastructureRef == nil {
-		// If the infrastructure ref is not set, marking the infrastructure as ready (no-op).
-		cluster.Status.InfrastructureReady = true
-		conditions.MarkTrue(cluster, clusterv1.InfrastructureReadyCondition)
+	cluster := s.cluster
+
+	// If the infrastructure ref is not set, no-op.
+	if s.cluster.Spec.InfrastructureRef == nil {
+		// if the cluster is not deleted, mark the infrastructure as ready to unblock other provisioning workflows.
+		if s.cluster.DeletionTimestamp.IsZero() {
+			cluster.Status.InfrastructureReady = true
+			conditions.MarkTrue(cluster, clusterv1.InfrastructureReadyCondition)
+		}
 		return ctrl.Result{}, nil
 	}
 
 	// Call generic external reconciler.
-	infraReconcileResult, err := r.reconcileExternal(ctx, cluster, cluster.Spec.InfrastructureRef)
+	obj, err := r.reconcileExternal(ctx, cluster, cluster.Spec.InfrastructureRef)
 	if err != nil {
-		return ctrl.Result{}, err
-	}
-	// Return early if we need to requeue.
-	if infraReconcileResult.RequeueAfter > 0 {
-		return ctrl.Result{RequeueAfter: infraReconcileResult.RequeueAfter}, nil
-	}
-	// If the external object is paused, return without any further processing.
-	if infraReconcileResult.Paused {
-		return ctrl.Result{}, nil
-	}
-	infraConfig := infraReconcileResult.Result
+		if apierrors.IsNotFound(err) {
+			s.infraClusterIsNotFound = true
 
-	// There's no need to go any further if the Cluster is marked for deletion.
-	if !infraConfig.GetDeletionTimestamp().IsZero() {
-		return ctrl.Result{}, nil
+			if !cluster.DeletionTimestamp.IsZero() {
+				// Tolerate infra cluster not found when the cluster is being deleted.
+				return ctrl.Result{}, nil
+			}
+
+			if cluster.Status.InfrastructureReady {
+				// Infra object went missing after the cluster was up and running
+				return ctrl.Result{}, reconcile.TerminalError(errors.Errorf("%s has been deleted after being ready", cluster.Spec.InfrastructureRef.Kind))
```

Comment:
Missed something on previous reviews. Let's assume we hit this branch + some other errors in other reconcile funcs.

The result will be that we don't get any requeue / exponential backoff.

But I guess when we hit other errors in addition to this one we still want to requeue?

Maybe we should just play it safe and not return any TerminalErrors in these reconcile phase funcs (also in other controllers)

Reply:
done!